pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -380,6 +381,14 @@ func (m *managerImpl) processShutdownEvent() error {
 					}
```

Comment:
just making sure I understand the existing code correctly:

Is kubelet marking the Pod as Failed even though it might still be terminating?

Reply:
Not really, there is a check here:
https://github.com/kubernetes/kubernetes/blob/3b6b8f9101d0d2086541dc06403e8d26ac63bbb1/pkg/kubelet/status/status_manager.go#L888
which gourds delay sending to API server the updated phase , reason and message until the containers are terminated. This can sometimes take a long if graceful shutdown for the node is specified.

Reply:
In that case, can we set the DisruptionTarget condition when the shutdown starts?

Reply:
Never mind, that could slow down the shutdown operations. Better just do it at the end. And if there is no chance to do it, the pod garbage collector could still set the condition once the kubelet is not responsive.

Reply:
Currently, the pod condition is added when shutdown starts. The `couldHaveRunningContainers` check only guards the change of the phase, reason and message. Meaning that the condition is added while the pod is in `Running` phase. I could delay adding of the condition as well, but it seems unnecessary and could risk not adding the condition at all if there is a kubelet restart in the meanwhile.

Reply:
I'll leave the decision to @bobbypage. Both approaches have their pros and cons.

Reply:
@bobbypage Can you comment on this since you are the author for this path?

Reply:
@mimowo's comment in https://github.com/kubernetes/kubernetes/pull/112360/files?show-viewed-files=true&file-filters%5B%5D=#r1013280124 is correct -- the shutdown manager sets the Failed phase as soon as the termination starts, however the status update of Failed terminal phase is not sent until the pod is terminal and all containers have terminated. This is necessary because the kubelet should only post terminal phase once all containers have terminated because otherwise the scheduler could schedule pods and the resources for the terminating pods could have not been released yet.

In the latest version of this PR, @mimowo updated to follow similar behavior -- the condition is only added after all the containers have terminated.